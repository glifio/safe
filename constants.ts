/* WALLET TYPES */
export const HD_WALLET = 'HD_WALLET'
export const SINGLE_KEY = 'SINGLE_KEY'

/* LOGIN OPTIONS */
export const IMPORT_MNEMONIC = 'IMPORT_MNEMONIC'
export const CREATE_MNEMONIC = 'CREATE_MNEMONIC'
export const IMPORT_SINGLE_KEY = 'IMPORT_SINGLE_KEY'

export const LEDGER = 'LEDGER'
export const METAMASK = 'METAMASK'

/* API ENDPOINTS */
export const FILSCAN = 'https://api.filscan.io:8700/v0/filscan'
export const FILSCAN_JSONRPC = 'https://api.filscan.io:8700/rpc/v1'
export const FILSCOUT = 'https://filscoutv3api.ipfsunion.cn'
export const FILFOX = 'https://filfox.info/api'

/* NETWORK VARS */
export const MAINNET = 'f'
export const TESTNET = 't'

export const GLIF_DISCORD = 'https://discord.gg/B9ju5Eu4Rq'
export const GLIF_TWITTER = 'https://twitter.com/glifio'

export const MAINNET_PATH_CODE = 461
export const TESTNET_PATH_CODE = 1

export const EXEC_ACTOR = 'f01'

/* FILECOIN APP VERSION MIN */
export const LEDGER_VERSION_MAJOR = 0
export const LEDGER_VERSION_MINOR = 18
export const LEDGER_VERSION_PATCH = 2

/* QUERY PARAMS */
export enum QPARAM {
  MSIG_ADDRESS = 'safe-id'
}

/* PAGES */
/* eslint-disable no-unused-vars */
export enum PAGE {
  SPEED_UP = '/speed-up',
  MSIG_HOME = '/home',
  MSIG_HISTORY = '/history',
  MSIG_PROPOSALS = '/proposals',
  MSIG_PROPOSAL = '/proposal',
  MSIG_ADMIN = '/admin',
  MSIG_WITHDRAW = '/withdraw',
  MSIG_APPROVE = '/approve',
  MSIG_CANCEL = '/cancel',
  MSIG_CHANGE_SIGNER = '/change-signer',
  MSIG_REMOVE_SIGNER = '/remove-signer',
  MSIG_ADD_SIGNER = '/add-signer',
  MSIG_CHANGE_APPROVALS = '/change-approvals',
  MSIG_CREATE_CONFIRM = '/create/confirm',
  MSIG_CREATE = '/create',
  MSIG_CHOOSE = '/choose',
  MSIG_CHOOSE_ACCOUNTS = '/accounts',
  CONNECT_LEDGER = '/connect/ledger',
  CONNECT_METAMASK = '/connect/metamask',
  NODE_DISCONNECTED = '/error/node-disconnected',
  LANDING = '/'
}
/* eslint-enable */
