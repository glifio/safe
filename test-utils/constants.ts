import { FilecoinNumber } from '@glif/filecoin-number'
import { getActorCode } from '@glif/filecoin-actor-utils'
// this breaks when import from regular react components
import { Network } from '../node_modules/@glif/react-components/dist/services/EnvironmentProvider'

export const WALLET_ADDRESS = 't1z225tguggx4onbauimqvxzutopzdr2m4s6z6wgi'
export const WALLET_ID = 't0101'
export const WALLET_ADDRESS_2 = 't1mbk7q6gm4rjlndfqw6f2vkfgqotres3fgicb2uq'
export const WALLET_ID_2 = 't0102'

// this is a premade multisig vesting actor
// if calibration net resets, these tests will fail
export const MULTISIG_ACTOR_ADDRESS =
  'f2m4f2dv7m35skytoqzsyrh5wqz3kxxfflxsha5za'
export const MULTISIG_ACTOR_ID = 't0201'

export const MULTISIG_SIGNER_ADDRESS = WALLET_ADDRESS
export const MULTISIG_SIGNER_ID = WALLET_ID
export const MULTISIG_SIGNER_ADDRESS_2 = WALLET_ADDRESS_2
export const MULTISIG_SIGNER_ID_2 = WALLET_ID_2

export const signers = [
  {
    robust: MULTISIG_SIGNER_ADDRESS,
    id: MULTISIG_SIGNER_ID
  },
  {
    robust: MULTISIG_SIGNER_ADDRESS_2,
    id: MULTISIG_SIGNER_ID_2
  }
]

export const mockStateGetActorRes = {
  Code: {
    '/': getActorCode('multisig', Network.CALIBRATION)
  },
  Balance: '80000000000'
}

const stateReadStateBaseRes = {
  Balance: new FilecoinNumber('1', 'fil').toAttoFil(),
  State: {
    InitialBalance: new FilecoinNumber('1', 'fil').toAttoFil(),
    NextTxnID: 2,
    NumApprovalsThreshold: 1,
    Signers: [],
    StartEpoch: 1000,
    UnlockDuration: 0
  }
}

export const mockStateReadStateSingleSignerRes = {
  ...stateReadStateBaseRes,
  State: {
    ...stateReadStateBaseRes.State,
    Signers: [MULTISIG_SIGNER_ID]
  }
}

export const mockStateReadStateDoubleSignerRes = {
  ...stateReadStateBaseRes,
  State: {
    ...stateReadStateBaseRes.State,
    Signers: [MULTISIG_SIGNER_ID, MULTISIG_SIGNER_ID_2]
  }
}
