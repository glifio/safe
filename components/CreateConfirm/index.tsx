import { useEffect, useMemo, useState } from 'react'
import { useRouter } from 'next/router'
import {
  getQueryParam,
  IconPending,
  Dialog,
  ShadowBox,
  ButtonRowCenter,
  ButtonV2Link,
  SmartLink,
  ErrorBox,
  WarningBox,
  useWallet,
  getAddrFromReceipt,
  useMessageReceiptQuery,
  useEnvironment
} from '@glif/react-components'

import { useMsig } from '../../MsigProvider'
import { PAGE, QPARAM } from '../../constants'

export const CreateConfirm = () => {
  const router = useRouter()
  const cid = getQueryParam.string(router, 'cid')
  const { setMsigActor, Address } = useMsig()
  const wallet = useWallet()
  const { coinType, explorerUrl } = useEnvironment()

  // Confirmation state
  const [creationError, setCreationError] = useState<boolean>(false)

  // Get the message receipt
  const { data: messageReceiptQuery, error: _messageReceiptError } =
    useMessageReceiptQuery({
      variables: { cid },
      pollInterval: 5000
    })

  const messageReceiptError = useMemo(() => {
    if (
      _messageReceiptError &&
      _messageReceiptError.message.includes("didn't find msg")
    ) {
      return null
    }
    return _messageReceiptError
  }, [_messageReceiptError])

  // Decode the receipt after receiving the message
  useEffect(() => {
    const receipt = messageReceiptQuery?.receipt
    if (!!receipt?.return) {
      try {
        // Verify the exit code
        if (receipt.exitCode !== 0) throw new Error('Safe creation failed')

        // Verify the safe address
        const { robust } = getAddrFromReceipt(receipt.return, coinType)
        if (!robust) throw new Error('Failed to get address from receipt')

        // Set the safe address
        setMsigActor(robust)
      } catch (e) {
        setCreationError(true)
      }
    }
  }, [messageReceiptQuery?.receipt, setMsigActor, coinType])

  // CID was not provided in query parameters
  if (!cid) {
    return (
      <Dialog>
        <ShadowBox>
          <h2>Safe creation status unknown</h2>
          <hr />
          <p>Missing CID for monitoring safe creation</p>
        </ShadowBox>
      </Dialog>
    )
  }

  // Something went wrong while retrieving the message
  if (messageReceiptError) {
    return (
      <Dialog>
        <ShadowBox>
          <h2>Safe creation status unknown</h2>
          <hr />
          <p>Something went wrong</p>
        </ShadowBox>
        <ErrorBox>
          <p>
            An error occurred while retrieving the status of your safe creation,
            please refresh the page.
          </p>
        </ErrorBox>
      </Dialog>
    )
  }

  // Something went wrong while creating the safe
  if (creationError) {
    return (
      <Dialog>
        <ShadowBox>
          <h2>Safe creation failed</h2>
          <hr />
          <p>Something went wrong</p>
        </ShadowBox>
        <ErrorBox>
          <p>
            An error occurred while creating your safe. Click on the CID below
            to view the transaction in the Glif Explorer:
          </p>
          <p>
            <SmartLink href={`${explorerUrl}/tx/${cid}`}>{cid}</SmartLink>
          </p>
        </ErrorBox>
      </Dialog>
    )
  }

  // The safe has been successfully created
  if (Address) {
    return (
      <Dialog>
        <ShadowBox>
          <h2>Safe creation completed</h2>
          <hr />
          <p>Your new Safe address is:</p>
          <p>
            <SmartLink href={`${explorerUrl}/address/${Address}`}>
              {Address}
            </SmartLink>
          </p>
        </ShadowBox>
        <WarningBox>
          Please make sure to store your Safe address before continuing.
        </WarningBox>
        <ButtonRowCenter>
          {wallet.robust || wallet.id ? (
            <ButtonV2Link large green href={PAGE.MSIG_HOME}>
              Go to the Safe dashboard
            </ButtonV2Link>
          ) : (
            <ButtonV2Link
              large
              green
              href={PAGE.LANDING}
              params={{
                [QPARAM.MSIG_ADDRESS]: Address
              }}
            >
              Connect your wallet to open the Safe
            </ButtonV2Link>
          )}
        </ButtonRowCenter>
      </Dialog>
    )
  }

  // The safe creation is in progress
  return (
    <Dialog>
      <ShadowBox>
        <h2>Safe creation in progress</h2>
        <hr />
        <IconPending height='2em' />
        <p>
          We&apos;re waiting for your transaction to confirm. Click on the
          transaction CID below to follow its progress in the Glif Explorer:
        </p>
        <p>
          <SmartLink href={`${explorerUrl}/tx/${cid}`}>{cid}</SmartLink>
        </p>
        <p>
          Your new Safe address will show here once the transaction confirms.
        </p>
      </ShadowBox>
    </Dialog>
  )
}
