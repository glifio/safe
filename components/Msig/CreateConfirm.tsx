import { useCallback, useEffect, useState } from 'react'
import { useRouter } from 'next/router'
import confirmMessage from '@glif/filecoin-message-confirmer'
import LotusRPCEngine from '@glif/filecoin-rpc-client'
import {
  IconPending,
  Dialog,
  ShadowBox,
  ButtonRowCenter,
  ButtonV2,
  SmartLink,
  ErrorBox
} from '@glif/react-components'

import { useMsig } from '../../MsigProvider'
import { PAGE } from '../../constants'
import { navigate } from '../../utils/urlParams'
import getAddrFromReceipt from '../../utils/getAddrFromReceipt'

export const CreateConfirm = () => {
  const [msigError, setMsigError] = useState(false)
  const [confirming, setConfirming] = useState(false)
  const { setMsigActor, Address } = useMsig()
  const router = useRouter()
  const { cid } = router.query
  const { NEXT_PUBLIC_EXPLORER_URL: explorerUrl } = process.env
  const { NEXT_PUBLIC_LOTUS_NODE_JSONRPC: apiAddress } = process.env

  const confirm = useCallback(async () => {
    try {
      const confirmed = await confirmMessage(cid as string, { apiAddress })
      if (confirmed) {
        const lCli = new LotusRPCEngine({ apiAddress })
        const receipt = await lCli.request<{
          ExitCode: number
          GasUsed: number
          Return: string
        }>('StateGetReceipt', { '/': cid }, null)

        // Verify exit code
        if (receipt.ExitCode !== 0) {
          setMsigError(true)
          return
        }

        // Verify address
        const { robust } = getAddrFromReceipt(receipt.Return)
        if (!robust) {
          setMsigError(true)
          return
        }

        // Set safe address
        setMsigActor(robust)
      }
    } catch (e) {
      setMsigError(true)
    }
  }, [cid, apiAddress, setMsigActor])

  useEffect(() => {
    if (!Address && !confirming && cid) {
      setConfirming(true)
      confirm()
    }
  }, [cid, Address, confirming, setConfirming, confirm])

  // CID was not provided in query parameters
  if (!cid) {
    return (
      <Dialog>
        <ShadowBox>
          <h2>Safe creation status unknown</h2>
          <hr />
          <p>Missing CID for monitoring safe creation</p>
        </ShadowBox>
      </Dialog>
    )
  }

  // Something went wrong while creating the safe
  if (msigError) {
    return (
      <Dialog>
        <ShadowBox>
          <h2>Safe creation failed</h2>
          <hr />
          <p>Something went wrong</p>
        </ShadowBox>
        <ErrorBox>
          <p>
            An error occurred while creating your safe. Click on the CID below
            to view the transaction in the Glif Explorer:
          </p>
          <p>
            <SmartLink href={`${explorerUrl}/message/?cid=${cid}`}>
              {cid}
            </SmartLink>
          </p>
        </ErrorBox>
      </Dialog>
    )
  }

  // The safe has been successfully created
  if (Address) {
    return (
      <Dialog>
        <ShadowBox>
          <h2>Safe creation completed</h2>
          <hr />
          <p>Your new Safe address is:</p>
          <p>
            <SmartLink href={`${explorerUrl}/actor/?address=${Address}`}>
              {Address}
            </SmartLink>
          </p>
        </ShadowBox>
        <ButtonRowCenter>
          <ButtonV2
            large
            green
            onClick={() => navigate(router, { pageUrl: PAGE.MSIG_HOME })}
          >
            Go to Safe dashboard
          </ButtonV2>
        </ButtonRowCenter>
      </Dialog>
    )
  }

  // The safe creation is in progress
  return (
    <Dialog>
      <ShadowBox>
        <h2>Safe creation in progress</h2>
        <hr />
        <IconPending flexShrink='0' />
        <p>
          We&apos;re waiting for your transaction to confirm. Click on the
          transaction CID below to follow its progress in the Glif Explorer:
        </p>
        <p>
          <SmartLink href={`${explorerUrl}/message/?cid=${cid}`}>
            {cid}
          </SmartLink>
        </p>
        <p>
          Your new Safe address will show here once the transaction confirms.
        </p>
      </ShadowBox>
    </Dialog>
  )
}
