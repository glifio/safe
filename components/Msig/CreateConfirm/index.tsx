import { useEffect, useState } from 'react'
import { useRouter } from 'next/router'
import LotusRPCEngine from '@glif/filecoin-rpc-client'
import {
  IconPending,
  Dialog,
  ShadowBox,
  ButtonRowCenter,
  ButtonV2,
  SmartLink,
  ErrorBox,
  WarningBox,
  navigate,
  useWallet,
  useMessageQuery
} from '@glif/react-components'

import { useMsig } from '../../../MsigProvider'
import { PAGE } from '../../../constants'
import { getAddrFromReceipt } from '../../../utils/getAddrFromReceipt'

interface Receipt {
  ExitCode: number
  GasUsed: number
  Return: string
}

const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms))

const recursivelyGetReceipt = async (
  apiAddress: string,
  cid: string,
  waitTime: number = 5000
): Promise<Receipt> => {
  await sleep(waitTime)
  const lCli = new LotusRPCEngine({ apiAddress })
  const receipt = await lCli.request<Receipt>(
    'StateGetReceipt',
    { '/': cid },
    null
  )

  if (receipt) return receipt

  // recursively, linearly ease off
  return recursivelyGetReceipt(apiAddress, cid, waitTime + 1000)
}

export const CreateConfirm = () => {
  const router = useRouter()
  const { cid } = router.query
  const { setMsigActor, Address } = useMsig()
  const wallet = useWallet()

  // Confirmation state
  const [fetchingReceipt, setFetchingReceipt] = useState<boolean>(false)
  const [receiptError, setReceiptError] = useState<boolean>(false)
  const [creationError, setCreationError] = useState<boolean>(false)

  // Get environment variables
  const { NEXT_PUBLIC_EXPLORER_URL: explorerUrl } = process.env
  const { NEXT_PUBLIC_LOTUS_NODE_JSONRPC: apiAddress } = process.env

  // Get the message by cid
  const { data: messageData, error: messageError } = useMessageQuery({
    variables: { cid: cid as string },
    pollInterval: 5000,
    skip: !cid
  })

  // Decode the receipt after receiving the message
  useEffect(() => {
    if (messageData?.message && !fetchingReceipt) {
      setFetchingReceipt(true)
      recursivelyGetReceipt(apiAddress, cid as string)
        .then((receipt) => {
          if (receipt) {
            try {
              // Verify the exit code
              if (receipt.ExitCode !== 0)
                throw new Error('Safe creation failed')

              // Verify the safe address
              const { robust } = getAddrFromReceipt(receipt.Return)
              if (!robust) throw new Error('Failed to get address from receipt')

              // Set the safe address
              setMsigActor(robust)
            } catch (e) {
              setCreationError(true)
            } finally {
              setFetchingReceipt(false)
            }
          }
        })
        .catch(() => setReceiptError(true))
    }
  }, [
    messageData?.message,
    apiAddress,
    cid,
    setMsigActor,
    fetchingReceipt,
    setFetchingReceipt
  ])

  // CID was not provided in query parameters
  if (!cid) {
    return (
      <Dialog>
        <ShadowBox>
          <h2>Safe creation status unknown</h2>
          <hr />
          <p>Missing CID for monitoring safe creation</p>
        </ShadowBox>
      </Dialog>
    )
  }

  // Something went wrong while retrieving the message or receipt
  if (messageError || receiptError) {
    return (
      <Dialog>
        <ShadowBox>
          <h2>Safe creation status unknown</h2>
          <hr />
          <p>Something went wrong</p>
        </ShadowBox>
        <ErrorBox>
          <p>
            An error occurred while retrieving the status of your safe creation,
            please refresh the page.
          </p>
        </ErrorBox>
      </Dialog>
    )
  }

  // Something went wrong while creating the safe
  if (creationError) {
    return (
      <Dialog>
        <ShadowBox>
          <h2>Safe creation failed</h2>
          <hr />
          <p>Something went wrong</p>
        </ShadowBox>
        <ErrorBox>
          <p>
            An error occurred while creating your safe. Click on the CID below
            to view the transaction in the Glif Explorer:
          </p>
          <p>
            <SmartLink href={`${explorerUrl}/message/?cid=${cid}`}>
              {cid}
            </SmartLink>
          </p>
        </ErrorBox>
      </Dialog>
    )
  }

  // The safe has been successfully created
  if (Address) {
    return (
      <Dialog>
        <ShadowBox>
          <h2>Safe creation completed</h2>
          <hr />
          <p>Your new Safe address is:</p>
          <p>
            <SmartLink href={`${explorerUrl}/actor/?address=${Address}`}>
              {Address}
            </SmartLink>
          </p>
        </ShadowBox>
        <WarningBox>
          Please make sure to store your Safe address before continuing to
          prevent losing access to your Safe.
        </WarningBox>
        <ButtonRowCenter>
          {wallet.address ? (
            <ButtonV2
              large
              green
              onClick={() => navigate(router, { pageUrl: PAGE.MSIG_HOME })}
            >
              Go to the Safe dashboard
            </ButtonV2>
          ) : (
            <ButtonV2
              large
              green
              onClick={() => navigate(router, { pageUrl: PAGE.LANDING })}
            >
              Connect your wallet to open the Safe
            </ButtonV2>
          )}
        </ButtonRowCenter>
      </Dialog>
    )
  }

  // The safe creation is in progress
  return (
    <Dialog>
      <ShadowBox>
        <h2>Safe creation in progress</h2>
        <hr />
        <IconPending flexShrink='0' />
        <p>
          We&apos;re waiting for your transaction to confirm. Click on the
          transaction CID below to follow its progress in the Glif Explorer:
        </p>
        <p>
          <SmartLink href={`${explorerUrl}/message/?cid=${cid}`}>
            {cid}
          </SmartLink>
        </p>
        <p>
          Your new Safe address will show here once the transaction confirms.
        </p>
      </ShadowBox>
    </Dialog>
  )
}
