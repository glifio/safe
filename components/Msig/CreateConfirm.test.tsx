import {
  act,
  getByRole,
  getByText,
  render,
  RenderResult
} from '@testing-library/react'
import { useMessageQuery } from '@glif/react-components'

import composeMockAppTree from '../../test-utils/composeMockAppTree'
import { CreateConfirm } from './CreateConfirm'
import { PAGE } from '../../constants'

jest.mock('../../MsigProvider')
jest.mock('@glif/react-components', () => {
  const original = jest.requireActual('@glif/react-components')
  return {
    ...original,
    useMessageQuery: jest.fn(({ variables }) => ({
      data: { message: { cid: variables.cid } }
    }))
  }
})

const mockStateGetMessageReceipt = jest.fn(async () => ({
  ExitCode: 0,
  Return: 'gkQAyogBVQJnC6HX7N9krE3QzLET9tDO1XuUqw==',
  GasUsed: 8555837
}))
jest
  .spyOn(require('@glif/filecoin-rpc-client'), 'default')
  .mockImplementation(() => {
    return {
      request: mockStateGetMessageReceipt
    }
  })

jest.spyOn(require('next/router'), 'useRouter').mockImplementation(() => {
  return {
    query: { cid: 'QmZ' },
    pathname: PAGE.MSIG_CREATE_CONFIRM,
    push: jest.fn()
  }
})

describe('confirmation of newly created multisig', () => {
  beforeEach(() => {
    jest.clearAllMocks()
    jest.useFakeTimers()
  })

  test('it renders message pending UI while the transaction is pending', async () => {
    const { Tree } = composeMockAppTree('pendingMsigCreate')
    let result: RenderResult | null = null

    await act(async () => {
      result = render(
        <Tree>
          <CreateConfirm />
        </Tree>
      )

      // we perform these tests inside act(), so we don't test things after they update
      const header = getByRole(result.container, 'heading')
      expect(header).toHaveTextContent('Safe creation in progress')
      expect(
        getByText(
          result.container,
          'Your new Safe address will show here once the transaction confirms.'
        )
      ).toBeInTheDocument()
      expect(result.container.firstChild).toMatchSnapshot()
    })
  })

  test('it attempts to confirm the pending msig create message', async () => {
    const { Tree } = composeMockAppTree('pendingMsigCreate')

    await act(async () => {
      render(
        <Tree>
          <CreateConfirm />
        </Tree>
      )
    })

    expect(useMessageQuery).toHaveBeenCalled()
  })

  test('it displays the multisig address after the message is confirmed', async () => {
    const { Tree } = composeMockAppTree('pendingMsigCreate')
    let result: RenderResult | null = null

    await act(async () => {
      result = render(
        <Tree>
          <CreateConfirm />
        </Tree>
      )
      jest.runOnlyPendingTimers()
    })

    const header = getByRole(result.container, 'heading')
    expect(header).toHaveTextContent('Safe creation completed')
    expect(useMessageQuery).toHaveBeenCalled()
    expect(mockStateGetMessageReceipt).toHaveBeenCalled()
    expect(result.container.firstChild).toMatchSnapshot()
  })
})
