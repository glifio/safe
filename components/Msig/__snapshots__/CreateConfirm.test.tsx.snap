// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`confirmation of newly created multisig it displays the multisig address after the message is confirmed 1`] = `
.c6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  gap: 0.75rem;
  padding: 0.5em 0.75em;
  color: var(--black);
  border-width: 1px;
  border-style: solid;
  border-radius: 8px;
  border-color: var(--black);
  background: transparent;
  font-size: 1.25em;
  color: var(--white);
  border-color: var(--green-medium);
  background: var(--green-medium);
}

.c6 svg path {
  stroke: var(--black);
}

.c6:hover {
  color: var(--purple-medium);
  border-color: var(--purple-medium);
}

.c6:hover svg path {
  stroke: var(--purple-medium);
}

.c6:active {
  color: var(--white) !important;
  border-color: var(--purple-medium) !important;
  background: var(--purple-medium) !important;
}

.c6:active svg path {
  stroke: var(--white) !important;
}

.c6 svg path {
  stroke: var(--white);
}

.c6:hover {
  color: var(--green-medium);
  border-color: var(--green-medium);
  background: var(--white);
}

.c6:hover svg path {
  stroke: var(--green-medium);
}

.c1 {
  padding: 1.5em;
  border-radius: 8px;
  overflow: hidden;
  text-align: center;
  word-break: break-word;
}

.c1 > *:first-child {
  margin-top: 0;
}

.c1 > *:last-child {
  margin-bottom: 0;
}

.c1 > svg:first-child {
  display: block;
  margin: 0 auto 1em;
}

.c1 > h2:first-child {
  margin: 0;
}

.c1 header:first-child {
  margin: -1.5em -1.5em 1.5em -1.5em;
  padding: 1.5em;
  background-color: var(--purple-medium);
  color: var(--white);
}

.c2 {
  background-color: var(--white);
  box-shadow: 0 0 0.5em var(--gray-light);
}

.c3 {
  background-color: var(--yellow-light);
  color: var(--yellow-dark);
}

.c3 hr {
  border-color: var(--yellow-dark);
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  margin: 1.5em 0;
}

.c5 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
  width: 100%;
  max-width: 35em;
}

.c0 form {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
}

<div
  class="c0"
>
  <div
    class="c1 c2"
  >
    <h2>
      Safe creation completed
    </h2>
    <hr />
    <p>
      Your new Safe address is:
    </p>
    <p>
      <a
        href="undefined/actor/?address=t2m4f2dv7m35skytoqzsyrh5wqz3kxxfflxsha5za"
        rel="noreferrer noopener"
        target="_blank"
      >
        t2m4f2dv7m35skytoqzsyrh5wqz3kxxfflxsha5za
      </a>
    </p>
  </div>
  <div
    class="c1 c3"
  >
    <svg
      fill="none"
      height="20px"
      viewBox="0 0 22 20"
      width="22px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M11 7.74982V9.99986M11 14.4999H11.0112M3.24282 19H18.7572C20.4815 19 21.5586 17.1246 20.6965 15.6249L12.9393 2.12474C12.0771 0.625088 9.92286 0.625088 9.0607 2.12474L1.30352 15.6249C0.441363 17.1246 1.5185 19 3.24282 19Z"
        stroke="var(--yellow-dark)"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      />
    </svg>
    Please make sure to store your Safe address before continuing to prevent losing access to your Safe.
  </div>
  <div
    class="c4 c5"
  >
    <button
      class="c6"
    >
      Connect your wallet to open the Safe
    </button>
  </div>
</div>
`;

exports[`confirmation of newly created multisig it renders message pending UI while the transaction is pending 1`] = `
.c3 {
  width: 24px;
  height: 24px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c4 {
  -webkit-animation: WNDHq 4s infinite;
  animation: WNDHq 4s infinite;
}

.c1 {
  padding: 1.5em;
  border-radius: 8px;
  overflow: hidden;
  text-align: center;
  word-break: break-word;
}

.c1 > *:first-child {
  margin-top: 0;
}

.c1 > *:last-child {
  margin-bottom: 0;
}

.c1 > svg:first-child {
  display: block;
  margin: 0 auto 1em;
}

.c1 > h2:first-child {
  margin: 0;
}

.c1 header:first-child {
  margin: -1.5em -1.5em 1.5em -1.5em;
  padding: 1.5em;
  background-color: var(--purple-medium);
  color: var(--white);
}

.c2 {
  background-color: var(--white);
  box-shadow: 0 0 0.5em var(--gray-light);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
  width: 100%;
  max-width: 35em;
}

.c0 form {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
}

<div
  class="c0"
>
  <div
    class="c1 c2"
  >
    <h2>
      Safe creation in progress
    </h2>
    <hr />
    <svg
      class="c3 c4"
      fill="none"
      height="24px"
      viewBox="0 0 24 24"
      width="24px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        fill="#FDB221"
        height="24"
        rx="12"
        width="24"
      />
      <mask
        fill="white"
        id="path-2-inside-1"
      >
        <path
          clip-rule="evenodd"
          d="M11.4006 11.9985L7.08534 6.60939C6.88876 6.36388 7.06355 6 7.37806 6H16.6265C16.9411 6 17.1159 6.36388 16.9193 6.60939L12.6017 12.0015L16.9169 17.3906C17.1135 17.6361 16.9387 18 16.6242 18H7.3757C7.06119 18 6.8864 17.6361 7.08298 17.3906L11.4006 11.9985Z"
          fill-rule="evenodd"
        />
      </mask>
      <path
        clip-rule="evenodd"
        d="M11.4006 11.9985L7.08534 6.60939C6.88876 6.36388 7.06355 6 7.37806 6H16.6265C16.9411 6 17.1159 6.36388 16.9193 6.60939L12.6017 12.0015L16.9169 17.3906C17.1135 17.6361 16.9387 18 16.6242 18H7.3757C7.06119 18 6.8864 17.6361 7.08298 17.3906L11.4006 11.9985Z"
        fill="#FDB221"
        fill-rule="evenodd"
      />
      <path
        d="M11.4006 11.9985L12.5715 12.9361L13.3222 11.9985L12.5715 11.061L11.4006 11.9985ZM7.08534 6.60939L5.91446 7.54695L5.91446 7.54695L7.08534 6.60939ZM16.9193 6.60939L15.7484 5.67183L15.7484 5.67183L16.9193 6.60939ZM12.6017 12.0015L11.4308 11.0639L10.6801 12.0015L11.4308 12.939L12.6017 12.0015ZM16.9169 17.3906L15.746 18.3282L15.746 18.3282L16.9169 17.3906ZM7.08298 17.3906L8.25387 18.3282L8.25387 18.3282L7.08298 17.3906ZM12.5715 11.061L8.25623 5.67183L5.91446 7.54695L10.2297 12.9361L12.5715 11.061ZM8.25623 5.67183C8.84599 6.40835 8.32162 7.5 7.37806 7.5V4.5C5.80547 4.5 4.93153 6.31941 5.91446 7.54695L8.25623 5.67183ZM7.37806 7.5H16.6265V4.5H7.37806V7.5ZM16.6265 7.5C15.683 7.5 15.1586 6.40836 15.7484 5.67183L18.0902 7.54695C19.0731 6.3194 18.1991 4.5 16.6265 4.5V7.5ZM15.7484 5.67183L11.4308 11.0639L13.7726 12.939L18.0902 7.54695L15.7484 5.67183ZM11.4308 12.939L15.746 18.3282L18.0878 16.453L13.7726 11.0639L11.4308 12.939ZM15.746 18.3282C15.1563 17.5916 15.6806 16.5 16.6242 16.5V19.5C18.1968 19.5 19.0707 17.6806 18.0878 16.453L15.746 18.3282ZM16.6242 16.5H7.3757V19.5H16.6242V16.5ZM7.3757 16.5C8.31925 16.5 8.84363 17.5916 8.25387 18.3282L5.9121 16.453C4.92916 17.6806 5.80312 19.5 7.3757 19.5V16.5ZM8.25387 18.3282L12.5715 12.9361L10.2297 11.061L5.9121 16.453L8.25387 18.3282Z"
        fill="#444444"
        mask="url(#path-2-inside-1)"
      />
    </svg>
    <p>
      We're waiting for your transaction to confirm. Click on the transaction CID below to follow its progress in the Glif Explorer:
    </p>
    <p>
      <a
        href="undefined/message/?cid=QmZ"
        rel="noreferrer noopener"
        target="_blank"
      >
        QmZ
      </a>
    </p>
    <p>
      Your new Safe address will show here once the transaction confirms.
    </p>
  </div>
</div>
`;
